### PR TITLE
Hw46 solution

### PR DESCRIPTION
**Pull request by OleksiiKolinko**

### CODE REVIEW

File: src/main/java/mate/academy/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+@RequestMapping(value = "/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping
+    @Operation(summary = "Get shopping cart",
+            description = "Get all information of  user shopping cart")
+    public ShoppingCartDto getShoppingCart(Authentication authentication, Pageable pageable) {
+        return shoppingCartService.getShoppingCart(authentication, pageable);
```
Comment (Rommelua) [inline src/main/java/mate/academy/controller/ShoppingCartController.java]: pass userId to Service layer (not  Authentication)

File: src/main/java/mate/academy/model/CartItem.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    private ShoppingCart shoppingCart;
+    @OneToOne(fetch = FetchType.LAZY)
```
Comment (Rommelua) [inline src/main/java/mate/academy/model/CartItem.java]: OneToOne?

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 3]
```diff
@@ -23,10 +27,19 @@ public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
     private final BookSpecificationBuilder bookSpecificationBuilder;
-    private final Set<Category> category;
+    private final CategoryRepository categoryRepository;
+    private final CategoryMapper categoryMapper;
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]: do not change BookServiceImpl class

File: src/main/java/mate/academy/repository/shoppingcart/ShoppingCartRepository.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.repository.shoppingcart;
+
+import mate.academy.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+
+    ShoppingCart findUserById(Long id);
```
Comment (Rommelua) [inline src/main/java/mate/academy/repository/shoppingcart/ShoppingCartRepository.java]: findUser?

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            Set<CartItem> cartItemSet = new HashSet<>();
+            cartItemSet.add(cartItem);
+            shoppingCart = new ShoppingCart();
+            shoppingCart.setUser(userRepository.findById(userId).orElseThrow(
+                    () -> new RegistrationException("User  with this id: "
+                            + userId + "is not exist")));
+            shoppingCart.setCartItems(cartItemSet);
+        } else {
+            shoppingCart.getCartItems().add(cartItem);
+        }
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
            addNewShoppingCart(userId);
        } 
        shoppingCart.getCartItems().add(cartItem);
```

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Authentication authentication, Pageable pageable) {
+        return shoppingCartMapper.toDto(shoppingCartRepository
+                .findUserById(getUserId(authentication)));
+    }
+
+    @Override
+    public CartItemDto addCartItem(CreateCartItemDto cartItemDto, Authentication authentication) {
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java]: @Transactional

File: src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        if (getMatch(cartItemId, authentication)) {
+            CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
+                    () -> new EntityNotFoundException("Can't find cart item with id: "
+                            + cartItemId));
+            ShoppingCart shoppingCart = shoppingCartRepository
+                    .findUserById(getUserId(authentication));
+            Set<CartItem> cartItems = shoppingCart.getCartItems();
+            cartItem.setQuantity(cartItemDto.quantity());
+            cartItems.add(cartItem);
+            shoppingCart.setCartItems(cartItems);
```
Comment (Rommelua) [inline src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
            ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
      .orElseThrow(EntityNotFoundException.supplier("Can't find shopping cart by id " + userId));
    CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, shoppingCart.getId())
      .orElseThrow(EntityNotFoundException.supplier(
        "Can't find cart item by id %s for user with id %s".formatted(cartItemId, userId))
      );
    cartItem.setQuantity(request.quantity());
```

File: src/main/java/mate/academy/controller/ShoppingCartController.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+@RequestMapping(value = "/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping
+    @Operation(summary = "Get shopping cart",
+            description = "Get all information of  user shopping cart")
+    public ShoppingCartDto getShoppingCart(Authentication authentication, Pageable pageable) {
+        return shoppingCartService.getShoppingCart(authentication.getName(), pageable);
```
Comment (koval-kn) [inline src/main/java/mate/academy/controller/ShoppingCartController.java]: you can access user in next way:
```suggestion
        User user = (User) authentication.getPrincipal();
        return shoppingCartService.getShoppingCart(user.getId(), pageable);
```

File: src/main/java/mate/academy/controller/ShoppingCartController.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping
+    @Operation(summary = "Add book to the shopping cart",
+            description = "Add book to the shopping cart")
+    public CartItemDto addCartItem(@RequestBody @Valid CreateCartItemDto cartItemDto,
+                                   Authentication authentication) {
+        return shoppingCartService.addCartItem(cartItemDto, authentication.getName());
```
Comment (koval-kn) [inline src/main/java/mate/academy/controller/ShoppingCartController.java]: pass only userId, like in prev comment

File: src/main/java/mate/academy/controller/ShoppingCartController.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+    @PreAuthorize("hasRole('USER')")
+    @PutMapping("/cart-items/{cartItemId}")
+    @Operation(summary = "Update quantity of a book in the shopping cart",
+            description = "Update quantity of a book in the shopping cart")
+    public CartItemDto updateQuantity(@PathVariable Long cartItemId,
+                               @RequestBody @Valid QuantityCartItemDto cartItemDto,
+                                      Authentication authentication) {
+        return shoppingCartService
+                .updateQuantity(cartItemId, cartItemDto, authentication.getName());
```
Comment (koval-kn) [inline src/main/java/mate/academy/controller/ShoppingCartController.java]: pass only userId, like in prev comment

File: src/main/java/mate/academy/controller/ShoppingCartController.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @DeleteMapping("/cart-items/{cartItemId}")
+    @Operation(summary = "Remove a book from the shopping cart",
+            description = "Remove a book from the shopping cart by id")
+    public void deleteCartItemId(@PathVariable Long cartItemId, Authentication authentication) {
+
+        shoppingCartService.deleteCartItemId(cartItemId, authentication.getName());
```
Comment (koval-kn) [inline src/main/java/mate/academy/controller/ShoppingCartController.java]: pass only userId, like in prev comment

**Review by koval-kn — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 12]
```diff
@@ -23,10 +24,12 @@ public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
     private final BookSpecificationBuilder bookSpecificationBuilder;
-    private final Set<Category> category;
 
     @Override
     public BookDto save(CreateBookRequestDto bookDto) {
+        Set<Long> categoryIdCreate = bookDto.getCategories().stream()
+                .map(CategoryDto::id)
+                .collect(Collectors.toSet());
```
Comment (Sanderfuck) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]: For what you need this set?

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Pull request merged**